cmake/kokkos_arch.cmake

```diff
@@ -574,15 +574,37 @@ CHECK_AMDGPU_ARCH(VEGA906 gfx906) # Radeon Instinct MI50 and MI60
 CHECK_AMDGPU_ARCH(VEGA908 gfx908) # Radeon Instinct MI100
 CHECK_AMDGPU_ARCH(VEGA90A gfx90a) # Radeon Instinct MI200
 
+FUNCTION(SET_AND_CHECK_AMD_ARCH ARCH FLAG)
+  SET(AMD_ARCHITECTURE ${ARCH} PARENT_SCOPE)
+  KOKKOS_SET_OPTION(ARCH_${ARCH} ON)
+  CHECK_AMDGPU_ARCH(${ARCH} ${FLAG})
+ENDFUNCTION()
+
 IF(KOKKOS_ENABLE_HIP AND NOT AMDGPU_ARCH_ALREADY_SPECIFIED)
   IF(KOKKOS_CXX_COMPILER_ID STREQUAL HIPCC)
     FIND_PROGRAM(ROCM_ENUMERATOR rocm_agent_enumerator)
     EXECUTE_PROCESS(COMMAND ${ROCM_ENUMERATOR} OUTPUT_VARIABLE GPU_ARCHS)
     STRING(LENGTH "${GPU_ARCHS}" len_str)
     # enumerator always output gfx000 as the first line
     IF(${len_str} LESS 8)
-      MESSAGE(SEND_ERROR "HIP enabled but no AMD GPU architecture currently enabled. "
-                         "Please enable one AMD GPU architecture via -DKokkos_ARCH_{..}=ON'.")
+      MESSAGE(SEND_ERROR "HIP enabled but no AMD GPU architecture could be automatically detected. "
+                         "Please manually enable one AMD GPU architecture via -DKokkos_ARCH_{..}=ON'.")
+    # check for known gpu archs, otherwise error out
+    ELSE()
+      STRING(REGEX MATCH "(gfx900|gfx906|gfx908|gfx90a)" DETECTED_GPU_ARCH ${GPU_ARCHS})
+      IF("${DETECTED_GPU_ARCH}" STREQUAL "gfx900")
+        SET_AND_CHECK_AMD_ARCH(VEGA900 ${DETECTED_GPU_ARCH})
```

Comment:
Won't this cause an error because `CHECK_AMDGPU_ARCH` is set multiple times (first on lines 572-575) I might be misunderstanding something.

Reply:
The autodetection is only entered if no user specified ARCH (lines 572-575) was detected as it is guarded by: `NOT AMDGPU_ARCH_ALREADY_SPECIFIED`